rfc/0023-kubewarden-stack-telemetry.md

```diff
@@ -0,0 +1,338 @@
+and the second creating a dedicated controller reconciler.
+
+> [!WARNING]
+> As a CNCF project, it is necessary to get approval from the CNCF to implement either of the following options.
+> [Telemetry Data Collection and Usage Policy | Linux Foundation](https://www.linuxfoundation.org/legal/telemetry-data-policy)
+
+## Option 1: OpenTelemetry Integration
+
+This implementation option aims for a less intrusive approach, decoupling the
+controller from the telemetry backend. With this design, all telemetry will be
+managed outside the Kubewarden controller's reconciliation loop. The controller
+will be instrumented with telemetry collection points that send data to an
+OpenTelemetry collector, which will then be configured to forward this data to
+a remote location.
+
+This option is inspired by previous observability work done in collaboration
+with SUSE.
+
+### Controller Changes
+
+The controller's `kubewarden-controller/internal/metrics`
+[package](https://github.com/kubewarden/kubewarden-controller/blob/main/internal/metrics/metrics.go)
+can be extended by adding new metrics to track the number of PolicyServers and
+policies deployed over time. These new metrics should also contain information
+about the Kubewarden and Kubernetes versions in use.
+
+The new metric can create time series for each policy like this, using the
+time-series database to group similar series:
+
+```go
+policiesGauge, err := meter.Int64Gauge("deployed-policies", metric.WithDescription("The number of deployed policies"))
+if err != nil {
+    log.Fatal(err)
+}
+// Attributes represent additional key-value descriptors that can be bound
+// to a metric observer or recorder.
+commonAttrs := []attribute.KeyValue{
+    attribute.String("policy_server_id", policy.GetPolicyServer().GetUID()),
+    attribute.String("policy_server_image", policy.GetPolicyServer().GetImageDigest()),
+    attribute.String("module", policy.GetModuleDigest()),
+    attribute.String("kubewarden_version", KUBEWARDEN_VERSION),
+    attribute.String("instance_uid", getInstanceUID()), // Corrected key name for consistency
+}
+
+policiesGauge.Record(context.Background(), 1, metric.WithAttributes(commonAttrs...)) // Corrected context.Background()
+```
+
+Once the metric is recorded, it is up to the OpenTelemetry collector to
+manipulate the data and send it to a remote location, which can be another
+OpenTelemetry collector or a time-series database (e.g., Prometheus).
+
+### OpenTelemetry Configuration
+
+In the OpenTelemetry collector configuration, a new pipeline should be added to
+redirect the new controller metric to a remote location for further analysis.
+This is an example of an initial configuration used to forward the metric to a
+remote OpenTelemetry collector.
+
+```yaml
+receivers:
+  otlp:
+    protocols:
+      grpc: {}
+processors:
+  # Control how much time we can wait before sending the metric to the remote Otel collector
+  batch:
+    timeout: 1h
```

Comment:
If we send the metrics every hour then it sounds more like real time monitoring

Reply:
This is a example configuration. It can be bigger. I just use the Longhorn [example](https://github.com/longhorn/upgrade-responder?tab=readme-ov-file#how-it-works). 

> Server will expect each instance of the application makes a request once per hour.